### ISSUE TITLE
Data Cube WMS Implementation

### ISSUE THREAD

**Issue by amruthkiran94** [opendatacube/datacube-core#444]
Greetings!

The [Datacube WMS](https://github.com/opendatacube/datacube-wms) git page doesn't seem to be accepting issues at the moment, posting here.

Is there a way to setup a WMS service using XAMPP/WAMP and the prescribed methods ( flask/mod-wsgi)?
Installation of mod-wsgi on XAMPP and Windows is excruciating. There has to be a simpler method?

**Comment by alexgleith** [comment]
Hey @amruthkiran94, the easiest way is to use the Docker images, available here: https://github.com/opendatacube/docker-images

The Dockerfile demonstrates a way to deploy the WMS implementation using Gunicorn. I haven't done a Windows Flask deployment, but I'm sure it's straight forward. Of course, you need ODC installed, so you kind of need a WOPI (Windows, ODC, Postgres and IIS!): https://medium.com/@bilalbayasut/deploying-python-web-app-flask-in-windows-server-iis-using-fastcgi-6c1873ae0ad8 

The M in your XAMPP is MySQL, right? You don't need that, if it is.

**Comment by alexgleith** [comment]
How are you going, @amruthkiran94? Does you need any more info on this issue?

**Comment by amruthkiran94** [comment]
Thank you @alexgleith for your quick reply. Sorry I could't get back to you on time.

I've tried out the Docker like you mention and it is completely new territory for me. Honestly, a few installation instructions on the page would be quite helpful.

I have an ODC environment with datasets ingested with proprietary data on Windows, due to the complication installing Flask/Mod-WSGI ( following the https://github.com/opendatacube/datacube-wms guide) I switched to Ubuntu, re-installed ODC and re-ingested), ran the database WMS query and ran the Docker code as stated in the link you kindly provided.

Although a local instance was running, the docker was still retrieving from the git and failed to install. Maybe I'm doing it wrong. I'll attach the screenshot soon.

**TL:DR:** Need better documentation on docker( installation, usage).

**Comment by alexgleith** [comment]
Hey @amruthkiran94, better documentation is a good idea. I'll work on putting something together.

For now, if you have a look at this Docker Compose file, this is enough to get a ODC instance running including WMS: https://github.com/opendatacube/docker-images/blob/master/wms/docker-compose.yml 

What you get from that is a Postgres database, a WMS server, including ODC installed, and Nginx as a reverse proxy. You can mount data into the WMS container, using a docker volume that is bound to a local folder, and then you can run your index/ingest process using that container too. There is a learning curve, but it's not too big!

Alternately, you can probably reverse-engineer the WMS Dockerfile there to see how to install and run Flask natively on Ubuntu.

**Comment by amruthkiran94** [comment]
Thank you @alexgleith, i'll try to work it out. There is a slight learning curve for beginners, But if the docker makes the overall process easier to carry out, it should IMHO be an integral part of ODC with refined documentation and examples. This would definitely help out beginners who rather use a 'ready-made' application than start the ODC from scratch.

**Comment by amruthkiran94** [comment]
@alexgleith This is the error I keep receiving when I try to run the Docker
![screenshot from 2018-05-22 15-11-07](https://user-images.githubusercontent.com/17108758/40354906-a7947a02-5dd2-11e8-926e-fd5ed1537e7d.png)

I have set the environment as follows 
![screenshot from 2018-05-22 15-16-15](https://user-images.githubusercontent.com/17108758/40355068-1b3f66b0-5dd3-11e8-8ad7-6aeef65015a7.png)

This is pointing to the installed ODC in Ubuntu, but running the docker from terminal causes the error. What am I doing wrong?

`sudo docker run  --rm opendatacube/wms gunicorn -b '0.0.0.0:8000' -w 5 --timeout 300 datacube_wms:wms
` 
This command runs the docker from the git retrieved while installation, how do i point it to the local instance?

**Comment by alexgleith** [comment]
Hey @amruthkiran94, I think the command needs to end in `datacube_wms.wms` that's the python path to the wsgi thing.

**Comment by alexgleith** [comment]
Might be worth you joining the Slack, if you're not on there already?

**Comment by amruthkiran94** [comment]
Thank you @alexgleith  looks like that was a typo in the documentation, yet I receive the same 
`Failed to find application object 'application' in 'datacube_wms.wms'
gunicorn.errors.HaltServer: <HaltServer 'App failed to load.' 4>
`.
Any suggestions?
Edit 1: is the [slack link](https://opendatacube.signup.team/) up? I'm not able to access it 

Edit 2: Tried using the Flask App method described in the WMS repo. Produces this error
![screenshot from 2018-05-22 16-21-27](https://user-images.githubusercontent.com/17108758/40358284-5296be0c-5ddc-11e8-90cd-c8e1426e200e.png)

**Comment by roarmstrong** [comment]
Hi @amruthkiran94, try using `gunicorn -b 0.0.0.0:8000 -w 5 --timeout 300 datacube_wms.wsgi` as the docker command.

The Flask App appears to have successfully started, if you perform the following query: `?service=WMS&request=GetCapabilities&version=1.3.0&tiled=true` you should see the WMS returning data based on the `wms_cfg.py` file

**Comment by alexgleith** [comment]
Sounds like it's running now, @amruthkiran94! 👏 

Next, as @roarmstrong says, is getting the right WMS URLs, and setting up your config.

As an aside, @roarmstrong, I think that the parameters in WMS should be case insensitive. GeoServer's implementation certainly is... I might put an issue up against the repo about it.

**Comment by amruthkiran94** [comment]
Thank you @roarmstrong  and @alexgleith  for your prompt replies.

Okay so I used the docker command `gunicorn -b 0.0.0.0:8000 -w 5 --timeout 300 datacube_wms.wsgi` but that still returned the `This XML file does not appear to have any style information associated with it` error like before.

How do I set URL correctly?  I used `http://0.0.0.0:8000/?service=WMS&request=GetCapabilities&version=1.3.0&tiled=true` but that returned a set of errors one of which states that `<ServiceException>
Unexpected server error: (psycopg2.OperationalError) fe_sendauth: no password supplied (Background on this error at: http://sqlalche.me/e/e3q8)
</ServiceException>` 
So I guess my environment variable isn't set properly?

**Comment by alexgleith** [comment]
That's not really an error, @amruthkiran94.

The URL is fine and the app is fine. Configuration may not be.

If you visit http://localhost:8000/?service=WMS&request=GetCapabilities&version=1.3.0 then it should work. (That's a basic WMS request.)

**Comment by amruthkiran94** [comment]
Configuration would be changed in the docker-compose.yml file correct ? what is the difference between `DB_DATABASE` and `POSTGRES_DB` @alexgleith ?
 I have a `Datacube_ubuntu` database within the `agdc` schema would it be the same?

**Comment by amruthkiran94** [comment]
So I was able to get the default WMS details from wms_config.py file as described by @roarmstrong. Screenshot attached below
![screenshot from 2018-05-23 15-21-07](https://user-images.githubusercontent.com/17108758/40417258-f67e8386-5e9c-11e8-89d0-798b2bbbda05.png)
This describes the example data, is there a format I need to follow to repeat for my datasets? 
Also, how do I view the map? QGIS/ArcGIS wms layer? ( Sorry I'm a Newbie )

**Comment by alexgleith** [comment]
Yep, well it's time that you went deeper, @amruthkiran94. We can't help you much more in a step-by-step manner. You'll need to just start going deep!

A hint for your next step is what @roarmstrong said:

> you should see the WMS returning data based on the wms_cfg.py file

And in terms of consuming the data, once it's returning actual data, you can use any app that supports WMS. That's a whole lesson in itself, but QGIS or ArcMap or Leaflet or... many other apps can consume WMSs.

**Comment by amruthkiran94** [comment]
Thank you so much for your kind help @alexgleith  and @roarmstrong . I have definitely learnt a lot!
I'll try customising the configuration file as per my datasets and see how that goes, but so far it has been a success!
Thanks again!
